File: src/common/py/data/entities/project_options.py
import typing
from dataclasses import dataclass, field

from dataclasses_json import dataclass_json

from .entity_types import ProjectFeatureID

UIOptions = typing.Dict[str, typing.Any]


@dataclass_json
@dataclass(frozen=True, kw_only=True)
class ProjectOptions:
    """
    Class holding all options of a **Project**.

    Attributes:
        optional_features: A list of all user-enabled optional features.
        ui: Arbitrary UI options.
    """

    optional_features: typing.List[ProjectFeatureID] = field(default_factory=list)

    ui: UIOptions = field(default_factory=dict)

File: src/common/web/data/entities/ProjectOptions.ts
import { Type } from "class-transformer";

import { type ProjectFeatureID } from "./EntityTypes";

/**
 * Arbitrary UI options.
 */
export type UIOptions = Record<string, any>;

/**
 * Class holding all options of a **Project**.
 *
 * @param optional_features - A list of all user-enabled optional features.
 * @param ui - Arbitrary UI options.
 */
export class ProjectOptions {
    // @ts-ignore
    @Type(() => String)
    public readonly optional_features: ProjectFeatureID[];

    public readonly ui: UIOptions;

    public constructor(optionalFeatures: ProjectFeatureID[] = [], uiOptions: UIOptions = {}) {
        this.optional_features = optionalFeatures;

        this.ui = uiOptions;
    }
}
